### PR TITLE
Added group `dns`

### PR DESCRIPTION
**Pull request by RichardLitt**
I was unable to get `recursive` working, and I'm not quite sure what it means or when it would be used. The example in the man pages seems to imply it is the default.

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -979,6 +979,130 @@ Outputs the list of peers that were removed.
...
+## dns [GET /dns{?arg}]
+Multihashes are hard to remember, but domain names are usually easy to
+remember. To create memorable aliases for multihashes, DNS TXT
+records can point to other DNS links, IPFS objects, IPNS keys, etc.
+This command resolves those links to the referenced object.
+
+The following response expamles assume this DNS TXT record:
+
+```
+ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
```
Comment (dignifiedquire) [inline apiary.apib]: Why is there a `.` after `ipfs.io`?

File: apiary.apib [hunk 2]
```diff
@@ -979,6 +979,130 @@ Outputs the list of peers that were removed.
...
+```
+
++ Parameters
+    + arg (string, required) - DNS name
+    + recursive (string, optional) - Resolve until the result is not a DNS link
```
Comment (dignifiedquire) [inline apiary.apib]: default value?

File: apiary.apib (hunk 1 above)
Comment (RichardLitt) [inline apiary.apib]: I don't know.

File: apiary.apib (hunk 2 above)
Comment (RichardLitt) [inline apiary.apib]: I don't know. @whyrusleeping? This is the relevant code:

```
    Run: func(req cmds.Request, res cmds.Response) {

        recursive, _, _ := req.Option("recursive").Bool()
        name := req.Arguments()[0]
        resolver := namesys.NewDNSResolver()

        depth := 1
        if recursive {
            depth = namesys.DefaultDepthLimit
        }
        output, err := resolver.ResolveN(req.Context(), name, depth)
        if err != nil {
            res.SetError(err, cmds.ErrNormal)
            return
        }
        res.SetOutput(&ResolvedPath{output})
    },
```

I can't tell if that is parsed by the CLI or not.

File: apiary.apib (hunk 1 above)
Comment (dignifiedquire) [inline apiary.apib]: @whyrusleeping @lgierth is this correct, or a typo?

**Comment by dignifiedquire** [comment]
Needs rebase

**Comment by RichardLitt** [comment]
Will get to that when closer to merge.

File: apiary.apib (hunk 1 above)
Comment (ghost) [inline apiary.apib]: It doesn't do harm, but I'd remove it. With CNAME records the record value has to end in `.` though :) E.g. `foobar.com CNAME gateway.ipfs.io.`.

Comment (RichardLitt) [inline apiary.apib]: Weird. This is what is in the go-ipfs docs; should I PR those?

Comment (daviddias) [inline apiary.apib]: Sure, probably that was just copy paste

**Comment by daviddias** [comment]
> I was unable to get recursive working, and I'm not quite sure what it means or when it would be used. 

It is to follow till it finds an IPFS hash in a subdomain of the domain being looked up. The help menu has an example:

>    For example, with this DNS TXT record:

```
  ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."

The resolver will give:

  > ipfs dns ipfs.io
  /ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy

And with this DNS TXT record:

  ipfs.ipfs.io. TXT "dnslink=/dns/ipfs.io ..."

The resolver will give:

  > ipfs dns ipfs.io
  /dns/ipfs.io
  > ipfs dns --recursive
  /ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy
```

File: apiary.apib (hunk 2 above)
Comment (RichardLitt) [inline apiary.apib]: 99% certain this means default is false.

**Comment by RichardLitt** [comment]
@diasdavid how can I handily test recursive to produce a reliable output for the http-api?

**Comment by daviddias** [comment]
@lgierth could you create a TXT record on ipfs.ipfs.io with `dnslink=/dns/ipfs.io` so that the example on the `ipfs dns` command works. Thank you :)

**Comment by dignifiedquire** [comment]
@RichardLitt needs rebase

**Comment by RichardLitt** [comment]
@dignifiedquire I know, thanks. Because of the way the master file is structured, each will need a rebase. If I haven't merged it, it's because I'm waiting for something from the assigned person.

File: apiary.apib [hunk 3]
```diff
@@ -1961,6 +1961,129 @@ Lists all available commands (and subcommands) and exits.
...
+## dns [GET /dns{?arg}]
+Multihashes are hard to remember, but domain names are usually easy to
+remember. To create memorable aliases for multihashes, DNS TXT
+records can point to other DNS links, IPFS objects, IPNS keys, etc.
+This command resolves those links to the referenced object.
+
+The following examples assume this DNS TXT record:
+
+```
+ipfs.io TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
```
Comment (ghost) [inline apiary.apib]: The hash later on is a different one

**Comment by ghost** [comment]
This LGTM with one change:
- [ ] hashes mismatch

The example works as is -- `/dns/<domain-name>/` is not a thing anymore, the old help is severly outdated in this regard.

Sorry for blocking this so long :/

File: apiary.apib (hunk 3 above)
Comment (ghost) [inline apiary.apib]: Maybe it'd be useful to use the `dig` command for DNS resolution? It'd give less experienced readers the tools they need in order to replicate what's being said. For example:

```
> dig +short TXT ipfs.io
dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy
```

Comment (RichardLitt) [inline apiary.apib]: I'm not sure. I think adding in CLI commands might be confusing?

**Comment by ghost** [comment]
LGTM :+1:

**Comment by RichardLitt** [comment]
Thanks @lgierth! And no worries, wasn't blocked long.

**Pull request merged**